internal/core/src/exec/HashTable.cpp

```diff
@@ -0,0 +1,272 @@
+        if (group_ && compare(group_, row_)) {
+            return group_;
+        }
+        const auto kEmptyGroup = BaseHashTable::TagVector::broadcast(0);
+        for (int64_t numProbedBuckets = 0;
+             numProbedBuckets < table.numBuckets();
+             ++numProbedBuckets) {
+            while (hits_ > 0) {
+                loadNextHit<op>(table);
+                if (compare(group_, row_)) {
+                    return group_;
+                }
+            }
+
+            uint16_t empty =
+                milvus::toBitMask(tagsInTable_ == kEmptyGroup) & kFullMask;
+            // if there are still empty slot available, try to insert into existing empty slot or tombstone slot
+            if (empty > 0) {
+                auto pos = milvus::bits::getAndClearLastSetBit(empty);
+                return insert(row_, bucketOffset_ + pos);
+            }
+            bucketOffset_ = table.nextBucketOffset(bucketOffset_);
+            tagsInTable_ = table.loadTags(bucketOffset_);
+            hits_ = milvus::toBitMask(tagsInTable_ == wantedTags_);
+        }
+        ThrowInfo(UnexpectedError,
+                  "Slots in hash table is not enough for hash operation, fail "
+                  "the request");
+    }
+
+ private:
+    static constexpr uint8_t kNotSet = 0xff;
+    template <Operation op, typename Table>
+    inline void
+    loadNextHit(Table& table) {
+        const int32_t hit = milvus::bits::getAndClearLastSetBit(hits_);
+        group_ = table.row(bucketOffset_, hit);
+        __builtin_prefetch(group_);
+    }
+
+    char* group_;
+    BaseHashTable::TagVector wantedTags_;
+    BaseHashTable::TagVector tagsInTable_;
+    int32_t row_;
+    int64_t bucketOffset_;
+    BaseHashTable::MaskType hits_;
+    //uint8_t indexInTags_ = kNotSet;
+};
+
+void
+HashTable::allocateTables(uint64_t size) {
+    AssertInfo(milvus::bits::isPowerOfTwo(size),
+               "Size:{} for allocating tables must be a power of two",
+               size);
+    AssertInfo(size > 0,
+               "Size:{} for allocating tables must be larger than zero",
+               size);
+    // Free existing table if present
+    if (table_ != nullptr) {
+        delete[] table_;
+        table_ = nullptr;
+    }
+    capacity_ = size;
+    const uint64_t byteSize = capacity_ * tableSlotSize();
+    AssertInfo(byteSize % kBucketSize == 0,
+               "byteSize:{} for hashTable must be a multiple of kBucketSize:{}",
+               byteSize,
+               kBucketSize);
+    numBuckets_ = byteSize / kBucketSize;
+    sizeMask_ = byteSize - 1;
+    sizeBits_ = __builtin_popcountll(sizeMask_);
+    bucketOffsetMask_ = sizeMask_ & ~(kBucketSize - 1);
+    // The total size is 8 bytes per slot, in groups of 16 slots with 16 bytes of
+    // tags and 16 * 6 bytes of pointers and a padding of 16 bytes to round up the
+    // cache line.
+    // TODO support memory pool here to avoid OOM
+    table_ = new char*[capacity_];
```

Comment:
table_ = static_cast<char*>(::operator new(byteSize, std::align_val_t(64)));

using char** seems to be dangerous.